pkg/network/usm/utils/connection_rollup.go

```diff
@@ -0,0 +1,284 @@
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+//go:build linux_bpf
+
+package utils
+
+import (
+	"github.com/DataDog/datadog-agent/pkg/network/types"
+)
+
+// ConnectionAggregator provides functionality for rolling-up datapoints from
+// different connections that refer to the same (client, server) pair
+type ConnectionAggregator struct {
+	data map[ipTuple]portValues
+}
+
+// NewConnectionAggregator returns a new instance of a `ConnectionAggregator`
+func NewConnectionAggregator() *ConnectionAggregator {
+	return &ConnectionAggregator{
+		data: make(map[ipTuple]portValues, 1000),
+	}
+}
+
+type ipTuple struct {
+	al uint64
+	ah uint64
+	bl uint64
+	bh uint64
+}
+
+type ephemeralPortSide uint8
+
+const (
+	ephemeralPortUnknown = iota
+	ephemeralPortA
+	ephemeralPortB
+)
+
+type portValues struct {
+	// these two values are populated by the *first* connection
+	// that has a given ipTuple
+	a uint16
+	b uint16
+
+	// this value is determined by analyzing the second connection that matches
+	// the associated ipTuple but has either a different port "a" or "b"
+	ephemeralSide ephemeralPortSide
+
+	// this is used to determine whether we should generate (a, b) or (b, a)
+	// when calling generateKey()
+	flipped bool
+}
+
+// RollupKey returns a _potentially_ modified key that is suitable for
+// aggregating datapoints belonging to the same (client, server) pair. On a
+// high-level, the function is supposed to return *one* key for all connections
+// matching (IP-A:*,IP-B:SERVER_PORT).
+//
+// This means that RollupKey(c1) == RollupKey(c2) == RollupKey(c3) for the
+// example below:
+//
+// c1: (IP-A:60001, IP-B:8080)
+// c2: (IP-A:60002, IP-B:8080)
+// c3: (IP-B:8080, IP-A:60003)
+//
+// The approach is very simplistic and only aims to address the common cases we
+// see in most workloads. The function will likely generate "false negatives",
+// but ideally it shouldn't generate "false positives". In other words, we may
+// not always rollup two different `types.ConnectionKey` that may legitimately
+// refer to the same (client, server) pair, but we shouldn't cause correctness
+// issues by aggregating two datapoints that *don't* belong to to the same
+// (client, server) pair.
+//
+// For the sake of documentation these are the edge-cases we *don't* intend to
+// address at this time:
+//
+// * Bare process monitoring (this code shouldn't even be reached when bare
+// process monitoring is enabled);
+// * Gossip/p2p protocols where you see symmetrical traffic like
+// (IP-A:PORT-X,IP-B:PORT-Y) and (IP-B:PORT-X,IP-A:PORT-Y);
+// * Containers running multiple processes instances bound to different ports. A
+// real world example we have here at Datadog are redis containers running
+// multiple instances. In this case only one one of the redis instances would
+// trigger a rollup.
```

Comment:
what will happen if both src port and dst port are small? 
for example 1234 and 2345?

or if the server port is ephemeral as well?

Reply:
The whole point about this approach is to completely ignore the actual port numbers (eg. are they large, do they belong to the ephemeral port range etc), and only determine what is client or server based on the "cardinality" of tuples.

> what will happen if both src port and dst port are small? for example 1234 and 2345?

If we have 

```
IP-A:1234 -> IP-B: 2345
``` 

We cant't conclude anything _yet_.

If we happen to see another connection like:
```
IP-A:1234 -> IP-B: 2346
```

Then we regard IP-B as being the client side, and will aggregate the 2 connections as

```
IP-A:1234 -> IP-B:0
```